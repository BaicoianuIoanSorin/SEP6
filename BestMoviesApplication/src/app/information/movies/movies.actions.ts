export class MoviesFetchNextPage {
  static readonly type = 'Movies - Fetch next page';
  constructor() {
  }
}

export class MoviesAddMovieToFavourites {
  static readonly type = 'Movies - Add movie to favourites';

  constructor(public userName: string,
              public movieId: number) {
  }
}

export class MoviesReset {
  static readonly type = 'Movies - Reset';
  constructor() {
  }
}

export class MoviesSearchTitle {
  static readonly type = 'Movies - Search by movie title';
  constructor(public movieTitle: string) {
  }
}

export class MoviesSearchReset {
  static readonly type = 'Movies - Search Reset';
  constructor() {
  }
}

