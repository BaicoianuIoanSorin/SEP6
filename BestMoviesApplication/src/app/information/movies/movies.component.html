<ng-container *ngIf="isFetching$ | async; then dataFetching; else dataFetched">
</ng-container>

<ng-template #dataFetched>
  <nb-card>
    <nb-card-header>
      <bm-search
        (searchEvent)="onSearch($event)"
        (resetSearch)="resetSearch()"
        [isFiltered]="isFiltered$ | async"
      ></bm-search>
    </nb-card-header>
    <nb-card-body>
      <ng-container *ngIf="isFiltered$ | async; then dataFiltered; else dataNotFiltered">
      </ng-container>
    </nb-card-body>
  </nb-card>
</ng-template>

<ng-template #dataFetching>
  <nb-spinner
    [nbSpinner]="true"
    status="basic"
    class="spinner">
  </nb-spinner>
</ng-template>

<ng-template #dataFiltered>
  <nb-list
    class="movie-container"
    *ngIf="allMovies$ | async as movies"
  >
    <nb-list-item *ngFor="let movie of movies">
      <nb-card size="medium" class="poster-movie">
        <nb-card-header class="header-poster-movie">
          <div class="movie-title">{{ movie.title }}</div>
        </nb-card-header>
        <nb-card-body class="body-poster-movie">
          <div class="show-more-button">
            <button
              nbButton
              status="primary"
              (click)="redirectToMovieOverviewPage(movie.id)"
            >
              <nb-icon icon="film-outline"></nb-icon>
              Show more
            </button>
          </div>
          <img *ngIf="movie.posterImage"
               [src]="movie.posterImage" [alt]="movie.title">
          <img *ngIf="!movie.posterImage"
               src="../../../assets/no-poster-available.jpg" [alt]="movie.title">
        </nb-card-body>
        <nb-card-footer class="footer-poster-movie">
          <div>
            <span class="movie-year">Year of publication: {{ movie.year }}</span>
          </div>
        </nb-card-footer>
      </nb-card>
    </nb-list-item>
  </nb-list>
</ng-template>

<ng-template #dataNotFiltered>
  <nb-list
    nbInfiniteList
    listenWindowScroll
    [threshold]="500"
    (bottomThreshold)="loadNext()"
    class="movie-container"
    *ngIf="moviesDisplayed$ | async as movies"
  >
    <nb-list-item *ngFor="let movie of movies">
      <nb-card size="medium" class="poster-movie">
        <nb-card-header class="header-poster-movie">
          <div class="movie-title">{{ movie.title }}</div>
        </nb-card-header>
        <nb-card-body class="body-poster-movie">
          <div class="show-more-button">
            <button
              nbButton
              status="primary"
              (click)="redirectToMovieOverviewPage(movie.id)"
            >
              <nb-icon icon="film-outline"></nb-icon>
              Show more
            </button>
          </div>
          <img *ngIf="movie.posterImage"
               [src]="movie.posterImage" [alt]="movie.title">
          <img *ngIf="!movie.posterImage"
               src="../../../assets/no-poster-available.jpg" [alt]="movie.title">
        </nb-card-body>
        <nb-card-footer class="footer-poster-movie">
          <div>
            <span class="movie-year">Year of publication: {{ movie.year }}</span>
          </div>
        </nb-card-footer>
      </nb-card>
    </nb-list-item>
  </nb-list>
</ng-template>
