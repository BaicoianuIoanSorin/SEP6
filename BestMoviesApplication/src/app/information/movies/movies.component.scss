.movie-information {

  .header-information {
    display: flex;
    justify-content: space-between;
  }

  .movie-container {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
    grid-gap: 1rem;
    max-width: 1200px;
    margin: 0 auto;

    .poster-movie {
      position: relative;
      width: 100%;
      border-radius: 15px 15px 15px 15px;
      box-shadow: 8px 6px 20px 2px #6d6d6d;

      .header-poster-movie {
        .movie-title {
          height: 2.8em;
        }
      }

      .body-poster-movie {
        padding: unset;
        overflow: hidden;
        position: relative;

        .show-more-button {
          position: absolute;
          top: 50%;
          left: 50%;
          transform: translate(-50%, -50%);
          opacity: 0;
          transition: opacity 0.3s ease;
          display: flex;
          align-items: center;
          justify-content: center;
          z-index: 1; /* Ensure the button is above the image */
        }

        img {
          width: 100%;
          height: 100%;
          object-fit: cover;
          object-position: center;
        }
      }

      .footer-poster-movie {
        margin-top: 0px;

        .person-birth {
          margin-top: 1rem;
        }
      }

    }

    .poster-movie:hover {
      opacity: 0.6;
      cursor: pointer;
    }

    .poster-movie:hover .body-poster-movie .show-more-button {
      opacity: 1;
    }

    .footer-page {
      display: flex;
      align-self: center;
    }
  }

  .movies {
    width: 100%;
  }

  .directors {
    width: 100%;

    nb-card-body {
      display: grid;
      gap: 0.4rem;
    }
  }
}

.spinner {
  color: nb-theme(color-primary-500);
}
