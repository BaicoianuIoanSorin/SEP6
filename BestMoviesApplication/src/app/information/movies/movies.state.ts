import {Movie} from "../../../model/movie";
import {Action, State, StateContext} from "@ngxs/store";
import {Injectable} from "@angular/core";
import {
  MoviesFetchInfo,
  MoviesReset, MoviesSearchReset, MoviesSearchTitle
} from "./movies.actions";
import {produce} from "immer";
import {moviesMock} from "../../../util/mocks/movies_mock";
import {NbToastrService} from "@nebular/theme";

export interface MoviesStateModel {
  isFetching: boolean;
  isFiltered: boolean;
  movies: Movie[];
}

export const defaultsState: MoviesStateModel = {
  isFetching: false,
  isFiltered: false,
  movies: [],
}

@State<MoviesStateModel>( {
  name: 'moviesPage',
  defaults: defaultsState,
})

@Injectable()
export class MoviesState {
  allMovies: Movie[] = [];

  constructor(
    private toastrService: NbToastrService
    //here the services used for getting date from backend are imported
  ) {
  }

  @Action(MoviesFetchInfo)
  async moviesFetchInfo(
    { getState, setState }: StateContext<MoviesStateModel>) {

    let currentState = getState();

    let newState = produce(currentState, draft => {
      draft.isFetching = true;
    })

    setState(newState);
    currentState = newState;

    // here we will call the API, but for now we have the mocks
    //TODO implement api call
    try {
      this.allMovies = moviesMock;
    }
    catch (e) {
      this.toastrService.show('danger', 'Fetching movies went wrong.');
      console.log(e);
    }

    newState = produce(currentState, draft => {
      draft.isFetching = false;
      //TODO instead of equal it with mocks, use the apis response
      draft.movies = this.allMovies;
    })

    setState(newState);
    currentState = newState;
  }

  @Action(MoviesSearchTitle)
  async moviesSearchTitle(
    { getState, setState }: StateContext<MoviesStateModel>,
    action: MoviesSearchTitle) {

    let currentState = getState();

    let newState = produce(currentState, draft => {
      draft.isFetching = true;
    })

    setState(newState);
    currentState = newState;

    //TODO implement api call or can be done also from frontend
    newState = produce(currentState, draft => {
      let movies = this.allMovies;
      if(movies.length < 0) {
        this.toastrService.show('', 'There are no movies to search');
      }
      else {
        let filteredMovies = movies.filter((movie: Movie) => {
          let value = movie.title.toLowerCase().includes(action.movieTitle);
          return value;
        });
        if(filteredMovies.length < 0) {
          this.toastrService.show('', 'There are no movies that contain ' + action.movieTitle);
        }
        draft.movies = filteredMovies;
      }
      draft.isFetching = false;
      draft.isFiltered = true;
    })

    setState(newState);
  }

  @Action(MoviesSearchReset)
  async moviesSearchReset(
    { getState, setState }: StateContext<MoviesStateModel>) {

    let currentState = getState();

    let newState = produce(currentState, draft => {
      draft.isFiltered = false;
      draft.movies = this.allMovies;
    })

    setState(newState);
  }

  @Action(MoviesReset)
  async moviesReset(
    { getState, setState }: StateContext<MoviesStateModel>) {
    setState(defaultsState);
  }
}
