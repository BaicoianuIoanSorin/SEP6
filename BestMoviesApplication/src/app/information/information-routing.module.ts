import {NgModule} from '@angular/core';
import {RouterModule, Routes} from '@angular/router';
import {NotFoundComponent} from "src/app/not-found/not-found.component";
import {MoviesComponent} from "src/app/information/movies/movies.component";

const routes: Routes = [
  {
    path: '',
    redirectTo: 'movies',
    pathMatch: 'full'
  },
  {
    path: 'movies',
    component: MoviesComponent,
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class InformationRoutingModule {
}
