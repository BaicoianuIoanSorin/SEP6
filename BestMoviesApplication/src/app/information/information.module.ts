import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import {
  NbLayoutModule,
  NbActionsModule,
  NbSidebarModule,
  NbMenuModule,
  NbCardModule,
  NbToastrModule,
  NbIconModule,
  NbButtonModule,
  NbSpinnerModule,
  NbListModule,
  NbSearchModule
} from '@nebular/theme';
import { NbEvaIconsModule } from '@nebular/eva-icons';
import { NgxsModule } from "@ngxs/store";
import { AgGridModule } from "ag-grid-angular";
import {MoviesState} from "src/app/information/movies/movies.state";
import {MoviesComponent} from "src/app/information/movies/movies.component";
import {InformationRoutingModule} from "src/app/information/information-routing.module";

export const states = [
  MoviesState,
];

@NgModule({
  declarations: [
    MoviesComponent
  ],
  imports: [
    CommonModule,
    InformationRoutingModule,
    NbMenuModule,
    NgxsModule.forFeature(states),
    NbSidebarModule,
    NbToastrModule,
    NbLayoutModule,
    NbButtonModule,
    NbEvaIconsModule,
    NbActionsModule,
    NbCardModule,
    NbIconModule,
    NbSpinnerModule,
    NbListModule,
    AgGridModule,
    NbSearchModule,
  ],
})
export class InformationModule { }
