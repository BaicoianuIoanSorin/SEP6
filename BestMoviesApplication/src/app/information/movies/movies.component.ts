import {Component, OnDestroy, OnInit} from '@angular/core';
import {NbSearchService} from "@nebular/theme";
import {
  OverAllInformationBestMoviesFetch,
  OverAllInformationFetchInfo, OverAllInformationReset
} from "src/app/overall-information/overall-information.actions";
import {CellClickedEvent, ColDef, ColumnApi, GridApi, GridReadyEvent} from "ag-grid-community";
import {Select, Store} from "@ngxs/store";
import {OverallInformationSelector} from "src/app/overall-information/overall-information.selector";
import {Observable} from "rxjs";
import {MoviesState} from "src/app/information/movies/movies.state";
import {MoviesSelector} from "src/app/information/movies/movies.selector";

@Component({
  selector: 'app-movies',
  templateUrl: './movies.component.html',
  styleUrls: ['./movies.component.scss']
})
export class MoviesComponent implements OnInit, OnDestroy{
  @Select(MoviesSelector.isFetching)
  isFetching$: Observable<boolean>;

  @Select(MoviesSelector.movies)
  movies$: Observable<boolean>;

  value = '';

  gridApi: GridApi;
  gridColumnApi: ColumnApi;

  alive: boolean = true;
  constructor(private store: Store,
              private searchService: NbSearchService) {

    this.searchService.onSearchSubmit()
      .subscribe((data: any) => {
        this.value = data.term;
      })

  }

  async ngOnInit() {
    const actionsInParallel = [
      new OverAllInformationFetchInfo(),
      new OverAllInformationBestMoviesFetch(),
    ];
    this.store.dispatch([...actionsInParallel]);
  }

  onGridReady(params: GridReadyEvent) {
    this.gridApi = params.api;
    this.gridColumnApi = params.columnApi;

    //TODO adjust autosize
    this.gridApi.sizeColumnsToFit();
  }

  onCellClicked( e: CellClickedEvent): void {
    // TODO redirect to the movie overview
  }

  columnDefs: ColDef[] = [
    {
      headerName: 'Id',
      field: 'id',
      valueFormatter: params => {
        console.log(params.value)
        return params.value;
      }
    },
    {
      headerName: 'Title',
      field: 'title',
    },
    {
      headerName: 'Year',
      field: 'year'
    }
  ]

  public defaultColDef: ColDef = {
    sortable: true,
    filter: true,
  };

  ngOnDestroy() {
    this.alive = false;
    this.store.dispatch(new OverAllInformationReset());
  }
}
