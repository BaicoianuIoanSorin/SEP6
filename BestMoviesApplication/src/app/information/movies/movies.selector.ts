import {Selector} from "@ngxs/store";
import {MoviesState, MoviesStateModel} from "./movies.state";

export class MoviesSelector {
  @Selector([MoviesState])
  static isFetching(state: MoviesStateModel) {
    return state.isFetching;
  }

  @Selector([MoviesState])
  static movies(state: MoviesStateModel) {
    return state.movies;
  }
}
