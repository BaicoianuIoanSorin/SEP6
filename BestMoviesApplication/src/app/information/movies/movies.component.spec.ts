import { ComponentFixture, TestBed } from '@angular/core/testing';
import {Spectator} from "@ngneat/spectator";
import {createComponentFactory, SpectatorFactory} from "@ngneat/spectator/jest";
import {
  NbActionsModule, NbButtonModule, NbCardModule,
  NbIconModule,
  NbLayoutModule, NbListModule,
  NbMenuModule, NbSearchModule,
  NbSidebarModule, NbSidebarService, NbSpinnerModule,
  NbToastrModule
} from "@nebular/theme";
import {NgxsModule} from "@ngxs/store";
import {NbEvaIconsModule} from "@nebular/eva-icons";
import {AgGridModule} from "ag-grid-angular";
import {MoviesComponent} from "./movies.component";
import {CommonModule} from "@angular/common";
import {InformationRoutingModule} from "../information-routing.module";
import {states} from "../information.module";

describe('app > information > movies > movies.component.spec.ts', () => {
  let component: MoviesComponent;
  let spectator: Spectator<MoviesComponent>;
  let createComponent: SpectatorFactory<MoviesComponent> = createComponentFactory({
    component: MoviesComponent,
    imports: [
      CommonModule,
      InformationRoutingModule,
      NbMenuModule,
      NgxsModule.forFeature(states),
      NbSidebarModule,
      NbToastrModule,
      NbLayoutModule,
      NbButtonModule,
      NbEvaIconsModule,
      NbActionsModule,
      NbCardModule,
      NbIconModule,
      NbSpinnerModule,
      NbListModule,
      AgGridModule,
      NbSearchModule,
    ],
    declarations: [],
    providers: [NbSidebarService],
  });

  beforeEach(() => {
    spectator = createComponent();
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('should match snapshot', async () => {
    spectator.detectChanges();
    await spectator.fixture.whenStable();
    spectator.detectChanges();
    await spectator.fixture.whenRenderingDone();

    // assert
    expect(spectator.fixture).toMatchSnapshot();
  });
});
