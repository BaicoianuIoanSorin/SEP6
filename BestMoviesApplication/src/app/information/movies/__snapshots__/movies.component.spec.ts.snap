// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app > information > movies > movies.component.spec.ts should match snapshot 1`] = `
<app-movies
  __isFetching$__selector={[Function Observable]}
  __isFiltered$__selector={[Function Observable]}
  __moviesDisplayed$__selector={[Function Observable]}
  nbToastrService={[Function NbToastrService]}
  placeHolder={[Function String]}
  router={[Function Router]}
  store={[Function Store]}
>
  <nb-card>
    <nb-card-header>
      <bm-search />
    </nb-card-header>
    <nb-card-body>
      <nb-list
        class="movie-container"
        listenwindowscroll=""
        nbinfinitelist=""
        role="list"
      >
        <nb-list-item
          role="listitem"
        >
          <nb-card
            class="poster-movie size-medium"
            size="medium"
          >
            <nb-card-header
              class="header-poster-movie"
            >
              <div
                class="movie-title"
              >
                La tierra de los toros
              </div>
            </nb-card-header>
            <nb-card-body
              class="body-poster-movie"
            >
              <div
                class="show-more-button"
              >
                <button
                  aria-disabled="false"
                  class="appearance-filled size-medium shape-rectangle icon-start status-primary"
                  nbbutton=""
                  status="primary"
                  tabindex="0"
                >
                  <nb-icon
                    icon="film-outline"
                  >
                    <svg
                      class="eva eva-film-outline"
                      fill="currentColor"
                      height="100%"
                      viewBox="0 0 24 24"
                      width="100%"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <g
                        data-name="Layer 2"
                      >
                        <g
                          data-name="film"
                        >
                          <rect
                            height="24"
                            opacity="0"
                            width="24"
                          />
                          <path
                            d="M18.26 3H5.74A2.74 2.74 0 0 0 3 5.74v12.52A2.74 2.74 0 0 0 5.74 21h12.52A2.74 2.74 0 0 0 21 18.26V5.74A2.74 2.74 0 0 0 18.26 3zM7 11H5V9h2zm-2 2h2v2H5zm4-8h6v14H9zm10 6h-2V9h2zm-2 2h2v2h-2zm2-7.26V7h-2V5h1.26a.74.74 0 0 1 .74.74zM5.74 5H7v2H5V5.74A.74.74 0 0 1 5.74 5zM5 18.26V17h2v2H5.74a.74.74 0 0 1-.74-.74zm14 0a.74.74 0 0 1-.74.74H17v-2h2z"
                          />
                        </g>
                      </g>
                    </svg>
                  </nb-icon>
                   Show more 
                </button>
              </div>
              <img
                alt="La tierra de los toros"
                src="https://image.tmdb.org/t/p/original/dB0CxRwLXJoOrRPUvsJjb2hnPtq.jpg"
              />
            </nb-card-body>
            <nb-card-footer
              class="footer-poster-movie"
            >
              <div>
                <span
                  class="movie-year"
                >
                  Year of publication: 2000
                </span>
              </div>
            </nb-card-footer>
          </nb-card>
        </nb-list-item>
      </nb-list>
    </nb-card-body>
  </nb-card>
</app-movies>
`;
